src/Agent.Worker/Release/ReleaseTrackingManager.cs

```diff
@@ -0,0 +1,200 @@
+            if (!File.Exists(file))
+            {
+                return null;
+            }
+
+            string content = File.ReadAllText(file);
+
+            return JsonConvert.DeserializeObject<ReleaseTrackingConfig>(content);
+        }
+
+        private void MarkForGarbageCollection(IExecutionContext executionContext, ReleaseTrackingConfig config)
+        {
+            Trace.Entering();
+
+            // Write a copy of the tracking config to the GC folder.
+            string gcDirectory = Path.Combine(
+                IOUtil.GetWorkPath(HostContext),
+                Constants.Release.Path.RootMappingDirectory,
+                Constants.Release.Path.GarbageCollectionDirectory);
+            string file = Path.Combine(
+                gcDirectory,
+                StringUtil.Format("{0}.json", Guid.NewGuid()));
+            WriteToFile(file, config);
+        }
+
+        public void MarkExpiredForGarbageCollection(IExecutionContext executionContext, TimeSpan expiration)
+        {
+            Trace.Entering();
+            Trace.Info("Scan all SourceFolder tracking files.");
+            string searchRoot = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Work), Constants.Release.Path.RootMappingDirectory);
+            if (!Directory.Exists(searchRoot))
+            {
+                executionContext.Output(StringUtil.Loc("GCDirNotExist", searchRoot));
+                return;
+            }
+
+            var allTrackingFiles = Directory.EnumerateFiles(searchRoot, Constants.Release.Path.TrackingConfigFile, SearchOption.AllDirectories);
+            Trace.Verbose($"Find {allTrackingFiles.Count()} tracking files.");
+
+            executionContext.Output(StringUtil.Loc("DirExpireLimit", expiration.TotalDays));
+            executionContext.Output(StringUtil.Loc("CurrentUTC", DateTime.UtcNow.ToString("o")));
+
+            // scan all sourcefolder tracking file, find which folder has never been used since UTC-expiration
+            // the scan and garbage discovery should be best effort.
+            // if the tracking file is in old format, just delete the folder since the first time the folder been use we will convert the tracking file to new format.
+            foreach (var trackingFile in allTrackingFiles)
+            {
+                try
+                {
+                    executionContext.Output(StringUtil.Loc("EvaluateTrackingFile", trackingFile));
+                    ReleaseTrackingConfig tracking = LoadIfExists(executionContext, trackingFile);
+                    
+                    if (tracking.LastRunOn == null)
+                    {
+                        Trace.Verbose($"{trackingFile} is a old format tracking file.");
+
+                        executionContext.Output(StringUtil.Loc("GCOldFormatTrackingFile", trackingFile));
+                        MarkForGarbageCollection(executionContext, tracking);
+                        IOUtil.DeleteFile(trackingFile);
```

Comment:
should we defer deleting tracking file until GC is complete? if not, we might not be able to delete folders which we could have done if next maintenance

Reply:
It will copy the tracking file in the GC folder and then delete it. Cleanup happens for all tracking files in the GC folder, if the maintenance couldn't finish, the tracking file will be picked up from GC folder when the next job runs